init.lua

```diff
@@ -261,7 +261,11 @@ local function rotate_body_and_head(player)
 		if #yaw_history > BODY_ROTATION_DELAY then
 			local body_yaw = table_remove(yaw_history, 1)
 			local player_yaw = player:get_look_horizontal()
-			return math_deg(player_yaw - body_yaw)
+			local angle = ((player_yaw - body_yaw + 3.0*math_pi) % (2.0*math_pi)) - math_pi
+			-- limit the body angle relative to head
+			local max = math_pi*0.3
```

Comment:
It's not smoothing. It's returning correct smallest angle for all values of angles.
If you look closely at how body rotates, it sometimes "flips" instead of correctly turning towards head direction.

Reply:
Sorry, this was written literally years ago, so I'm just trying to remember now.
I think, this may only matter when we're trying to clamp the angle (second part of this commit - min/max thing).
If you just pass raw yaw numbers, it will return correct absolute rotation, but not the "shortest" one, and thus clamping will be incorrect.
![player_angles](https://github.com/minetest-mods/playeranim/assets/138972279/be1b8ea6-a0a8-44e8-820b-a08478b0d8c6)

Reply:
I implemented [similar code in Minetest](https://github.com/minetest/minetest/blob/4acbd59162/src/util/numeric.h#L434-L451) that moves into the shorter path, except with a threshold of 180°. If I understand it correctly, you're using 54° (0.3 * PI) here to decide whether to pick the short path?
I'm asking this because servers can be somewhat laggy, thus going strictly by the shorter path (diverge from 180° difference) might be even more reliable?

In either case, I'd very much appreciate if you could add comments to the code explaining *what exactly* it does. The next person who's looking at it might have similar issues at understanding.

Reply:
This change was made after I saw players standing with their heads turned backwards (or in other unnatural neck rotations) during lag. :)

```lua
local max = math_pi*0.3
```
Is an arbitrary limit that feels "good enough" - can be changed to anything from 0 to below 180.
This will be the maximum angle head can be rotated relative to the body.
Normal IRL necks can turn more, but I tweaked this angle to be less, because during lag, 70° or more did not look good.

```lua
local short_angle = ((angle + 3.0*math_pi) % (2.0*math_pi)) - math_pi
```

This is just a general formula for shortest rotation around circle. It's possible to use a bunch of `if`s and `abs()` and `sign()` or whatever, to choose the shortest path, but this code is just a branchless version in pure math.
The idea of how it works is:
We can map any input angle (from -360° to +360°) into positive 0-360° range by adding 360° (`2*pi`) to it and using modulo operation. But since we want the shortest rotation, we need half of those values become negative. So before doing modulo, we add 180° (another `pi`, so `3*pi` total) and after doing modulo, it's subtracted it back. This "normalizes" any weird input angle and maps this into [-180°,180) range. After this is done, we can just limit it to the range we want.

Actually, originally I wanted to not just limit, but try to animate body turning with different speeds depending on the difference in angles, so this is the reason for getting correct shortest angle first...

I will add comments to code in a bit too.